azure/service.ts

```diff
@@ -21,23 +21,241 @@ import { RunError } from "@pulumi/pulumi/errors";
+        name: string,
+        props: cloud.HostProperties | undefined,
+        containers: cloud.Containers): GroupsInfo {
+
+    props = props || {};
+    const groups: azure.containerservice.Group[] = [];
+
+    let firstContainerName: string | undefined;
+    let firstContainerPort: number | undefined;
+    const exposedPorts: ExposedPorts = {};
+
+    for (const containerName of Object.keys(containers)) {
+        const container = containers[containerName];
+
+        if (firstContainerName === undefined) {
+            firstContainerName = containerName;
+        }
+
+        const ports = container.ports;
+        let targetPortNumber: number | undefined;
+        let protocol: string | undefined;
+        let isPublic: boolean | undefined;
+        if (ports) {
+            if (ports.length >= 2) {
+                throw new RunError("Only zero or one port can be provided with a container: " + containerName);
+            }
+
+            if (ports.length === 1) {
+                const port = ports[0];
+                targetPortNumber = port.targetPort !== undefined ? port.targetPort : port.port;
+                protocol = port.protocol;
+
+                if (port.port !== undefined && targetPortNumber !== port.port) {
+                    throw new RunError("Mapping a host port to a different target port is not supported in Azure currently.");
+                }
+
+                if (containerName === firstContainerName) {
+                    firstContainerPort = targetPortNumber;
+                }
+
+                const external = port.external === undefined ? false : port.external;
+                if (isPublic === undefined) {
+                    // first port we're seeing.  set the isPublic value based on that.
+                    isPublic = external;
+                }
+                else if (isPublic !== external) {
+                    // have an existing port.  Values have to match.
+                    throw new RunError("All ports must have a matching [external] value.");
+                }
+            }
+        }
+
+        if (isPublic === false) {
+            throw new RunError("Only public ip address types are supported by Azure currently.");
+        }
+
+        const imageName = getImageName(container);
+        if (!imageName) {
+            throw new Error("[getImageName] should have always produced an image name.");
+        }
+
+        const registry = container.build ? getOrCreateGlobalRegistry() : undefined;
+
+        const imageOptions = computeImage(this, imageName, container, registry);
+
+        const memoryInGB = pulumi.output(container.memoryReservation).apply(
+            r => r === undefined ? 1 : r / 1024);
+
+        const disallowedChars = /[^-a-zA-Z0-9]/g;
+        const qualifiedName = (name + "-" + containerName).replace(disallowedChars, "-");
+
+        const azureContainer: AzureContainer = {
+            name: qualifiedName,
+            cpu: 1,
```

Comment:
should this be `cpu` from the container definition?

Reply:
We actually don't support specifying the number of CPUs at the container definition level.  @lukehoban is that intentional?

Reply:
Not intentional. I left a comment on this PR suggesting you add this support as part of this PR since you are touching several of the relevant places. It was an oversight that this parameter wasn’t originally added to cloud.Container.

Reply:
Opened https://github.com/pulumi/pulumi-cloud/pull/568